Add a demo backend with compiler

iseeyuan opened the pull request:
This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52870 [Lite Interpreter] Support features from to_backend
* **#52603 Add a demo backend with compiler**

Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

iseeyuan committed:
Add a demo backend with compiler

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"




[ghstack-poisoned]

raziel commented on test/cpp/jit/CMakeLists.txt:
```diff
@@ -74,7 +78,7 @@ if(USE_SYSTEM_ONNX)
 endif()
 
-set(JIT_TEST_DEPENDENCIES jitbackend_test gtest)
+set(JIT_TEST_DEPENDENCIES torch gtest jitbackend_test backend_with_compiler)
```
torch wasn't needed before?

iseeyuan replied:
torch is included into jitbackend_test and backend_with_compiler, so not necessarily needed here. It's just to explicitly show the dependency.

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,111 @@
+// This test JIT backend is intended to do the minimal amount of work
+// necessary to test that the JIT backend registration endpoints and
+// code generation are working correctly. It is not intended to
+// produce numerically correct results.
```
Maybe move this doc up there (see my other comment), so it's clear it covers both preprocess and the backend interface.

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,111 @@
+namespace jit {
+
+namespace {
+std::vector<std::string> parseBlob(const std::string& blob) {
```
Since there could be many "blobs", maybe rename "parseBlob" to something more specific (e.g. parseMethodHandle or something like that).

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,111 @@
+  c10::impl::GenericDict compile(
+      c10::IValue processed,
+      c10::impl::GenericDict method_compile_spec) override {
+    return processed.toGenericDict();
```
This is the next TODO right? Actually receive +1 methods and split them here.

iseeyuan replied:
```processed``` here is a dict with method names as keys. So it's already split? I was thinking of to move the parsing part here: to parse to blob to something that can be directly used in execute.

raziel replied:
You're right, that makes sense.

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,111 @@
+
+// Actual AOT compilation function. Put here for demonstration of backend
+// as a whole piece. It's used when compilation is required. A dummy function
+// can be passed when there's no usage of compilation in runtime backend lib.
```
I'd just clarify: "there's no usage of AOT compilation " since "compilation" could still theoretically take place on-target.

I'll clarify/add the docs to these interfaces.

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,111 @@
+// The output of this process would produce a dictionary
+// Key: method name.
+// Val: compiled blob (represented by a string).
+c10:
```
Fix format.

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,111 @@
+
+namespace torch {
+namespace jit {
+
```
It'd be good to document up here the behaviour/contract of this backend: e.g. Implementation of a PyTorch Backend that can process, compile and execute TorchScript Modules composed of 'add' and 'sub' operators between constants and expect 2 inputs of type Tensor of blah... which are added or subtracted to blah blah

You can punt this until you add support for constants.

But basically it'd be good if this backend, however simple (it really doesn't need to be fancy), implemented some well defined algorithm that used all the bits the preprocessed/compiled Module rather than just hardcoding some values, and ignoring others.

iseeyuan replied:
Good suggestion! I'll add more comments, and discuss with you on defined algorithms offline.

iseeyuan replied:
Now I get your point of "well defined algorithm". I'll add the support for constant.

raziel commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -32,9 +33,97 @@ TEST(BackendTest, ToBackend) {
   AT_ASSERT(res[1].toTensor().equal(ref[1].toTensor()));
 }
+
+TEST(BackendTest, TestCompiler) {
```
Just double checking, "TestCompiler" and "ToBackend" test the same. The only difference is that "TestCompiler" is simpler (single output vs tuple), correct?

iseeyuan replied:
The major difference is their backends. ```ToBackend``` tests the "test_backend" backend, which does not have an actual compiler but everything is hard-coded. ```TestCompiler``` is testing a new backend, "backend_with_compiler_demo", to test the compilation of a "supported" feature.

raziel replied:
Got it. I missed the different backend class.
Maybe once you're done with the new dummy backend we don't need the other one.

iseeyuan replied:
Good point. I need to have the python frontend test for this backend as well. After that is done we could remove the other one.

raziel commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -32,9 +33,97 @@ TEST(BackendTest, ToBackend) {
+  compile_spec.insert("forward", fake_dict);
+  auto any_dict_ty = DictType::create(StringType::get(), AnyType::get());
+  // lowered module
+  //  auto lm = torch::jit::detail::codegen_backend_module(
```
delete commented code?

raziel reviewed (commented):
Thanks Martin, this is great.

I left some minor comments in the code.

The main suggestion, which is something you're probably already working on, is to make the backend really implement execution capabilities according to what was compiled rather than hardcoding some constants and ignoring those in the original Module (nothing fancy though).

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend pass through the AOT compiled blob. (TODO: parsing the blob to the format that the backend can understand can happen here.)
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 




[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend pass through the AOT compiled blob. (TODO: parsing the blob to the format that the backend can understand can happen here.)
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

kimishpatel commented on test/cpp/jit/backend_with_compiler.cpp:
`const auto&`

kimishpatel commented on test/cpp/jit/backend_with_compiler.cpp:
const..

kimishpatel commented on test/cpp/jit/backend_with_compiler.cpp:
By the way aten::add vs aten::add_. The runtime actually executes in place version rather than aten::add.

iseeyuan replied:
Good to know. Thanks!

SplitInfinity commented on test/cpp/jit/backend_with_compiler.cpp:
```suggestion
// For this backend, the actual compilation happens in preprocess function AOT.
```

SplitInfinity reviewed (approved):
I left a few comments inline (mostly spelling errors).

Maybe it's okay for a demo backend, but it seems weird to preprocess a module by creating a list of `std::string` instructions. Why not encode the instructions in some way?

raziel commented on test/cpp/jit/backend_with_compiler.cpp:
```diff
@@ -0,0 +1,159 @@
+            instruction);
+        auto sub = instruction.substr(15);
+        const_val = stod(sub);
+      } else if (token == "aten::add") {
+        output_list.emplace_back(x.add_(h, const_val));
+      } else if (token == "aten::sub") {
+        output_list.emplace_back(x.sub_(h, const_val));
```
As discussed offline, update the docs above to describe what this backend actually implements, which I think it's just support for Modules that implement a sum or subtraction of 2 inputs (i.e. in1 + in2 or in1 - in2)

iseeyuan replied:
Thanks! Comment added and submitting this PR soon, after resolving CI.

raziel reviewed (approved):
Thanks!

iseeyuan commented on test/cpp/jit/test_utils.h:
```diff
@@ -5,6 +5,14 @@
+  try {                                                                  \
+    (void)statement;                                                     \
+    ASSERT_TRUE(false);                                                  \
+  } catch (const std::exception& e) {                                    \
```
The purpose of this macro is to compare the exception message itself to differentiate the exact throw. It's used to check expected messages from all exception types.

iseeyuan commented on test/cpp/jit/test_backend.cpp:
```diff
@@ -32,9 +33,95 @@ TEST(BackendTest, ToBackend) {
       "test_backend", m, compile_spec, any_dict_ty);
+  // lowered module code:
+  /*
+class test_backendLoweredModule(Module):
```
Thanks. Fixed.

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]

iseeyuan committed:
Update on "Add a demo backend with compiler"

This PR introduced a backend with minimum compilation capability to the to_<backend> flow. The targets are:

- Demonstrate the end-to-end flow with adding a backend -> compilation -> runtime
- How the backend compilation errors be surfaced to the user, with the original model's source code information. (C++ only in this PR. Python APIs will be demonstrated in a following PR.)

Changes:

- Compilation

1. A backend with minimum compilation features, "backend_with_compiler_demo" is added.
2. The compilation happens AOT in the ```pre_process``` function registered to this backend.
3. Compiled results are stored in a string blob for each method. They are serialized to the lowered module with ```__get_state__``` function.
4. Error message with model source code is thrown, for features not handled by the backend compiler.


- Runtime

1. The compiled blob is loaded in ```__set_state__``` method.
2. The ```compile``` function of the backend parses the blob to the format that the backend can understand.
3. The ```execute``` function of the backend executes the specified method (handle).

Tests: 

- ```BackendTest.TestCompiler```: the C++ end-to-end demonstration on a supported model. After compilation and running, the lowered model produces the same result as the original torchscript model.
- ```BackendTest.TestCompilerNotSupport```: Demonstrate the error message from the AOT compilation for a feature not supported from the input module. The error message looks like:

```
"The node of aten::mul is not supported in this compiler. Source code:   File "<string>", line 3

    def forward(self, x, h):
        return x * h
               ~~~~~ <--- HERE
```

 


Differential Revision: [D26593968](https://our.internmc.facebook.com/intern/diff/D26593968)

[ghstack-poisoned]
